### PR TITLE
Add configuration for setting the Jenkins URL.

### PR DESCRIPTION
**Pull request by tkuhlman**
This change adds the public_url charm configuration option so you can
specify the Jenkins URL and prefix.

### CODE REVIEW

**Comment by ryan-beisner** [comment]
There should be corresponding unit and amulet test coverage for the new config option.  This reactive rewrite has pretty good unit and amulet test coverage and I'd really like to see us keep that level of trust as contributions are made.  Thank you.

**Comment by tkuhlman** [comment]
There are corresponding unit tests in my submission.

As far as amulet tests, I couldn't get them to work with Juju2 and I have no Juju1 environments anymore. In my dev instead I used an mojo spec to test the full integration, I was hoping Travis would exercise the amulet tests for me, but I see that isn't the case. Is there a place where I can run amulet test any changes or a trick to getting them to work with Juju2? The bug I hit is https://bugs.launchpad.net/juju/+bug/1608723

The basic amulet tests will of course use the default config value, I would just add changing it and verifying the change was made.

**Comment by freeekanayaka** [comment]
Tim Kuhlman <notifications@github.com> writes:

> There are corresponding unit tests in my submission.
>
> As far as amulet tests, I couldn't get them to work with Juju2 and I
> have no Juju1 environments anymore.

They used to pass on Juju 2. Something has probably regressed, I'll have
a look tomorrow.

> In my dev instead I used an mojo
> spec to test the full integration, I was hoping Travis would exercise
> the amulet tests for me, but I see that isn't the case.

Yes, this a known pain point. There is some on going work from the
juju-solutions team to provide easy glue between Jenkins and GitHub in
order to run charm integration tests on custom Jenkins servers (Travis
is unfortunately to limited to run juju integration tests properly).

> Is there a
> place where I can run amulet test any changes or a trick to getting
> them to work with Juju2? The bug I hit is
> https://bugs.launchpad.net/juju/+bug/1608723

Good question. The situation has gone a bit wild since Juju 2 AFAIK,
what I can say is that I use bundletester.

Remember however that integration tests are supposed to be
self-contained executables, so provided you have the required Python
packages in your PATH and that you have a clean Juju 2 model at hand,
you can just invoke:

./tests/NN_some_test
  
> The basic amulet tests will of course use the default config value, I
> would just add changing it and verifying the change was made.

Sounds good.

**Comment by tkuhlman** [comment]
Thanks! I was able to get the amulet tests working with Juju2 by invoking the amulet tests directly which obviously I missed before. I'll add in some amulet tests later today.

File: .gitignore [hunk 1]
```diff
@@ -6,3 +6,4 @@
 *.pyc
 lib/charms/__init__.py
 lib/charms/layer/__init__.py
+.idea
```
Comment (freeekanayaka) [inline .gitignore]: This looks like a leftover?

File: config.yaml [hunk 2]
```diff
@@ -48,3 +48,9 @@ options:
     default: 1
     description: |
       Number of executors to configure for jenkins master.
+  public_url:
```
Comment (freeekanayaka) [inline config.yaml]: Please use dashes as separator, like the other config keys in this file:

"public_url" -> "public-url"

File: config.yaml [hunk 3]
```diff
@@ -48,3 +48,9 @@ options:
...
+    type: string
+    default: ""
+    description: >
```
Comment (freeekanayaka) [inline config.yaml]: Please use a pipe here (for consistencies with other config keys in this file): 

"description: >" -> "description: |"

File: lib/charms/layer/jenkins/api.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 import requests
+from urllib.parse import urlparse
```
Comment (freeekanayaka) [inline lib/charms/layer/jenkins/api.py]: Please follow [PEP8](https://www.python.org/dev/peps/pep-0008/#imports) conventions, placing absolute imports before relative imports, and standard library imports before third-party and local library imports (all groups separated by a blank line):

```python
import requests

import jenkins

from urllib.parse import urlparse

from charmhelpers.core.hookenv import ERROR

from charms.layer.jenkins.credentials import Credentials
```

File: lib/charms/layer/jenkins/api.py [hunk 5]
```diff
@@ -34,13 +34,21 @@
 class Api(object):
     """Encapsulate operations on the Jenkins master."""
 
+    @property
+    def url(self):
+        config = hookenv.config()
+        prefix = urlparse(config["public_url"]).path
+        if not prefix:
+            prefix = "/"
+        return "http://localhost:8080{}".format(prefix)
```
Comment (freeekanayaka) [inline lib/charms/layer/jenkins/api.py]: Please take advantage of the URL constants defined in charms.layer.jenkins.service and use urljoin:

```python
return urljoin(URL, perfix)
```

File: lib/charms/layer/jenkins/configuration.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 import os
+from urllib.parse import urlparse
```
Comment (freeekanayaka) [inline lib/charms/layer/jenkins/configuration.py]: Add a blank line before this import, as per PEP8.

File: lib/charms/layer/jenkins/configuration.py [hunk 7]
```diff
@@ -11,8 +12,47 @@
 class Configuration(object):
     """Manage global Jenkins configuration."""
 
+    @staticmethod
+    def _set_prefix(prefix):
```
Comment (freeekanayaka) [inline lib/charms/layer/jenkins/configuration.py]: Please move private methods at the bottom, after public ones, since people reading the code are typically interested in the public APIs, at least at first.

File: lib/charms/layer/jenkins/configuration.py [hunk 8]
```diff
@@ -11,8 +12,47 @@
...
+        if update:
+            with open(paths.DEFAULTS_CONFIG_FILE, 'w') as defaults_file:
+                defaults_file.write(defaults_content + "\n")
+            return True
+
+        return False
+
     def bootstrap(self):
-        """Generate Jenkins' initial config."""
+        """Generate Jenkins' initial config. """
```
Comment (freeekanayaka) [inline lib/charms/layer/jenkins/configuration.py]: There's a leftover blank space at the end of the comment.

File: lib/charms/layer/jenkins/configuration.py [hunk 9]
```diff
@@ -11,8 +12,47 @@
 class Configuration(object):
     """Manage global Jenkins configuration."""
 
+    @staticmethod
```
Comment (freeekanayaka) [inline lib/charms/layer/jenkins/configuration.py]: There's no need for the @staticmethod decorator, please just add "self" to the signature and make it a regular instance method.

File: tests/basic.py [hunk 10]
```diff
@@ -129,6 +130,28 @@ def test_10_change_password(self):
         self.spec.deployment.configure(charm_name, {"password": PASSWORD})
         self.spec.deployment.sentry.wait()
 
+    def test_10_change_url(self):
+        """Validate that after changing public_url we can login at the new
+            prefix and that the public_url has been updated.
+        """
```
Comment (freeekanayaka) [inline tests/basic.py]: Please format the docstring like this:

```python
 """
Validate that after changing public_url we can login at the new
 prefix and that the public_url has been updated.
 """
```

File: unit_tests/test_configuration.py [hunk 11]
```diff
@@ -34,6 +36,46 @@ def test_bootstrap(self):
             FileContains(matcher=Contains("<numExecutors>1</numExecutors>")))
         self.assertEqual({8080}, self.fakes.juju.ports["TCP"])
 
+    def test_set_prefix(self):
+        test_file = tempfile.NamedTemporaryFile()
```
Comment (freeekanayaka) [inline unit_tests/test_configuration.py]: This file will be left around after the test suite completes. Please use something like:

```
from fixtures import TempDir

# ...

    test_file = self.useFixture(TempDir()).join("test-file")
```

See the [fixtures](https://pypi.python.org/pypi/fixtures) package for more details.

File: unit_tests/test_configuration.py [hunk 12]
```diff
@@ -34,6 +36,46 @@ def test_bootstrap(self):
...
+
+        # Case #4 - previous config, different prefix, expected change
+        updated = self.configuration._set_prefix("/jenkins-alt")
+        self.assertTrue(updated)
+
+        # Case #5 - previous config, no prefix, expected change
+        updated = self.configuration._set_prefix("")
+        self.assertTrue(updated)
+
+        # Case #6 - no config file, no expected change
```
Comment (freeekanayaka) [inline unit_tests/test_configuration.py]: Please break down all these separate cases in separate test methods. You'll probably want a small helper function to create the test file.

In this way if something breaks you can have an instant look at all edge cases that don't pass, instead of having to run the suite N times and discover a new failure each time.

File: unit_tests/test_users.py [hunk 13]
```diff
@@ -33,6 +33,10 @@ def test_configure_admin_custom_password(self):
         If a password is provided, it's used to configure the admin user.
         """
         self.fakes.juju.config["password"] = "x"
+        # TODO figure out why self.fakes.juju setting is not reliably working.
```
Comment (freeekanayaka) [inline unit_tests/test_users.py]: Can you clarify this comment? I'll try to given an explanation or maybe fix an bug in the underlying machinery.

File: .gitignore (hunk 1 above)
Comment (tkuhlman) [inline .gitignore]: Out of habit I added that in since I use pycharm, I'll remove it.

File: lib/charms/layer/jenkins/api.py (hunk 5 above)
Comment (tkuhlman) [inline lib/charms/layer/jenkins/api.py]: There is no URL constant in service anymore, I removed it. A constant is no longer valid with the ability to change prefix and it was only used by API calls so it makes more sense to be in the api file rather than in the service file. This method is the replacement for the constant.

File: lib/charms/layer/jenkins/configuration.py (hunk 6 above)
Comment (tkuhlman) [inline lib/charms/layer/jenkins/configuration.py]: That does conform to pep8, urllib is part of the standard library and it is grouped with os also part of the standard library with relative imports coming second.

File: unit_tests/test_configuration.py (hunk 11 above)
Comment (tkuhlman) [inline unit_tests/test_configuration.py]: No NamedTemporaryFile defaults to deleting the file when it is closed so it doesn't hang around after the test suite completes. I'm happy to use the fixtures though.

File: unit_tests/test_users.py (hunk 13 above)
Comment (tkuhlman) [inline unit_tests/test_users.py]: The problem here is I expected from looking at other code that self.fakes.juju.config["password"] would set the config option for the password but in my testing it did nothing it wasn't until I used hookenv.config() to get the config and set it that way that the test begun to work.

File: unit_tests/test_configuration.py (hunk 11 above)
Comment (freeekanayaka) [inline unit_tests/test_configuration.py]: Yeah, I should have mentioned that the case were it hangs around is if the test method fails. In that case the close() call never gets executed (an exception is raised before).

**Comment by tkuhlman** [comment]
I made updates as above, squashed all the commits together and rebased. I didn't end up switching NamedTemporaryFile for a fixture because there is only a temporary dir for the fixture so I still have to open a file an write to it anyway.

File: unit_tests/test_configuration.py [hunk 14]
```diff
@@ -42,3 +91,19 @@ def test_migrate(self):
             fd.write("")
         self.configuration.migrate()
         self.assertThat(paths.LEGACY_BOOTSTRAP_FLAG, Not(FileExists()))
+
+    @contextlib.contextmanager
+    def _temporary_defaults_file(self):
+        """ Setup a temporary DEFAULTS_CONFIG_FILE
+        :return: a temporary file with the name attribute being the path
+        """
+        test_file_path = self.useFixture(fixtures.TempDir()).join("test-file")
```
Comment (freeekanayaka) [inline unit_tests/test_configuration.py]: Thanks for using the TempDir fixture, so no temporary files are left around when the test fails.

However, I just realized that I could have suggested you a way simpler approach with systemfixtures:

http://systemfixtures.readthedocs.io/en/latest/?badge=latest#filesystem

that's the pattern that the rest of the unit test code in this charm uses. In your case I think you would have to write something like:

```python
self.fakes.fs.add(paths.DEFAULTS_CONFIG_FILE)
os.makedirs("/etc/default")
```
and from this point any call to open("/etc/default/jenkins") will be magically redirected to a temporary file (something like "/tmp/tmpxxx/etc/default/jenkins).

You could add the code above to unit_tests.states.AptInstallJenkins, and since ConfigurationTest already uses the AptInstalledJenkins state fixtures all you unit tests will have that default file already setup (since it's a file that is always expected to be there in a real unit).

I realize the above is a tad tricky, if you have troubles getting it done let me know, or just land this PR as it is and I'll take care of this small clean up myself.

**Comment by tkuhlman** [comment]
Great that does clean up the tests nicely. I rebased and squashed to a single commit again since a new change just merged.

**Comment by tkuhlman** [comment]
Ah timing, I missed the fix from PR #16, which I am glad to see merge, one more rebase coming up.

**Comment by freeekanayaka** [comment]
Thanks, the change looks good to me. Do you have commit rights or should I trigger the merge myself?

**Comment by tkuhlman** [comment]
I don't have commit rights.
